Sources/XcodeProjMapper/Mappers/Graph/XcodeGraphMapper.swift

```diff
@@ -0,0 +1,291 @@
+            let projectPath = path.appending(component: projectName)
+            let xcodeProj = try XcodeProj(pathString: projectPath.pathString)
+            return .project(xcodeProj)
+        }
+
+        throw XcodeGraphMapperError.noProjectsFound(path.pathString)
+    }
+
+    /// Builds the final `Graph` by enumerating the `.xcodeproj` or `.xcworkspace`.
+    ///
+    /// If it's a workspace, we gather all `.xcodeproj` references. If it's a single project,
+    /// we treat it like a workspace with a single project. Then we load each project,
+    /// map packages & targets, and build dependency edges. This merges logic from your old `GraphMapper`.
+    func buildGraph(from graphType: XcodeMapperGraphType) async throws -> Graph {
+        // A place to store discovered: path -> Project, packages, dependencies, etc.
+        var projects: [AbsolutePath: Project] = [:]
+        var packages: [AbsolutePath: [String: Package]] = [:]
+        var dependencies: [GraphDependency: Set<GraphDependency>] = [:]
+        var dependencyConditions: [GraphEdge: PlatformCondition] = [:]
+
+        // Step 1: Identify all .xcodeproj paths in "the workspace" or the single .xcodeproj
+        let projectPaths: [AbsolutePath]
+        let graphName: String
+        let workspacePath: AbsolutePath
+
+        switch graphType {
+        case let .workspace(xcworkspace):
+            workspacePath = try xcworkspace.pathOrThrow
+            graphName = workspacePath.basenameWithoutExt
+
+            // Gather all .xcodeproj references in the workspace
+            let projectRefs = try await extractProjectPaths(
+                from: xcworkspace.data.children,
+                srcPath: workspacePath.parentDirectory
+            )
+            projectPaths = projectRefs.isEmpty ? [] : projectRefs
+
+        case let .project(xcodeProj):
+            let projPath = try xcodeProj.pathOrThrow
+            workspacePath = projPath.parentDirectory
+            graphName = "Workspace"
+            // For a single .xcodeproj, treat it like a workspace with a single project
+            projectPaths = [projPath]
+        }
+
+        // Step 2: Build a synthetic "Workspace" model for the final Graph
+        let workspace = Workspace(
+            path: workspacePath,
+            xcWorkspacePath: workspacePath,
+            name: graphName,
+            projects: projectPaths
+        )
+
+        // Step 3: For each project path, load the .xcodeproj and map to a `Project`
+        let projectResults = try workspace.projects.map { path -> (AbsolutePath, Project) in
+            // Instead of "ProjectProvider", just load XcodeProj directly
+            let xcodeProj = try XcodeProj(pathString: path.pathString)
+            let projectMapper = PBXProjectMapper()
+            let project = try projectMapper.map(xcodeProj: xcodeProj)
+            return (path, project)
+        }
```

Comment:
What do you think about turning some of these more complex steps into helper methods where the name of the method would clearly indicate what it does?